File: ext/types/container_id.cpp
#include <iostream>
#include "types.h"

using namespace mesos;

ContainerIDPtr toContainerID(char* bs, int len)
{
  ContainerIDPtr val = new ContainerID();
  val->set_value(bs, len);
  return val;
}

int fromContainerID(ContainerIDPtr p, char** poke)
{
  *poke = (char*) p->mutable_value()->data();
  return p->mutable_value()->size();
}

void destroyContainerID(ContainerIDPtr p)
{
  delete p;
}

File: src/System/Mesos/Raw/ContainerId.hs
module System.Mesos.Raw.ContainerId where
import           System.Mesos.Internal

type ContainerIDPtr = Ptr ContainerID

foreign import ccall unsafe "ext/types.h toContainerID" c_toContainerID :: ToID ContainerIDPtr

foreign import ccall unsafe "ext/types.h fromContainerID" c_fromContainerID :: FromID ContainerIDPtr

foreign import ccall unsafe "ext/types.h destroyContainerID" c_destroyContainerID :: ContainerIDPtr -> IO ()

instance CPPValue ContainerID where

  marshal x = do
    (strp, l) <- cstring $ containerIDId' x
    liftIO $ c_toContainerID strp $ fromIntegral l

  unmarshal p = fmap ContainerID $ do
    ptrPtr <- alloc
    len <- liftIO $ c_fromContainerID p ptrPtr
    peekCString' (ptrPtr, len)

  destroy = c_destroyContainerID
